transport/grpc_test.go

```diff
@@ -0,0 +1,273 @@
+	}, func(t *testing.T, testEnv *testEnv) {
+		request, err := newTestJSONRequest("example", "Foo::Bar", &testBarRequest{One: "hello"})
+		require.NoError(t, err)
+		response, err := testEnv.Transport.Call(context.Background(), request)
+		require.NoError(t, err)
+		require.NotNil(t, response)
+		testBarResponse := &testBarResponse{}
+		require.NoError(t, json.Unmarshal(response.Body, testBarResponse))
+		require.Equal(t, "hello", testBarResponse.One)
+	})
+}
+
+func TestGRPCError(t *testing.T) {
+	doWithTestEnv(t, "example-caller", 5, []transport.Procedure{
+		newTestJSONProcedure("example", "Foo::Bar", testBar),
+	}, func(t *testing.T, testEnv *testEnv) {
+		request, err := newTestJSONRequest("example", "Foo::Bar", &testBarRequest{Error: "hello"})
+		require.NoError(t, err)
+		_, err = testEnv.Transport.Call(context.Background(), request)
+		require.Equal(t, yarpcerrors.UnknownErrorf("hello"), err)
+	})
+}
+
+type testBarRequest struct {
+	One   string
+	Error string
+}
+
+type testBarResponse struct {
+	One string
+}
+
+func testBar(ctx context.Context, request *testBarRequest) (*testBarResponse, error) {
+	if request == nil {
+		return nil, nil
+	}
+	if request.Error != "" {
+		return nil, errors.New(request.Error)
+	}
+	return &testBarResponse{
+		One: request.One,
+	}, nil
+}
+
+func newTestJSONRequest(service string, method string, request interface{}) (*Request, error) {
+	body, err := json.Marshal(request)
+	if err != nil {
+		return nil, err
+	}
+	return &Request{
+		TargetService: service,
+		Method:        method,
+		Body:          body,
+	}, nil
+}
+
+func newTestJSONProcedure(service string, name string, handler interface{}) transport.Procedure {
+	procedure := yarpcjson.Procedure(name, handler)[0]
+	procedure.Service = service
+	return procedure
+}
+
+func doWithTestEnv(
+	t *testing.T,
+	caller string,
+	numInbounds int,
+	procedures []transport.Procedure,
+	f func(*testing.T, *testEnv),
+) {
+	testEnv, err := newTestEnv(caller, numInbounds, procedures)
+	require.NoError(t, err)
+	defer func() {
+		assert.NoError(t, testEnv.Close())
+	}()
+	f(t, testEnv)
+}
+
+type testEnv struct {
+	Caller         string
+	Transport      TransportCloser
+	YARPCTransport *grpc.Transport
+	YARPCInbounds  []*grpc.Inbound
+}
+
+func newTestEnv(
+	caller string,
+	numInbounds int,
+	procedures []transport.Procedure,
+) (_ *testEnv, err error) {
+	yarpcTransport := grpc.NewTransport()
+	if err := yarpcTransport.Start(); err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err != nil {
+			err = multierr.Append(err, yarpcTransport.Stop())
+		}
+	}()
+
+	addresses := make([]string, numInbounds)
+	yarpcInbounds := make([]*grpc.Inbound, numInbounds)
+	for i := 0; i < numInbounds; i++ {
+		listener, err := net.Listen("tcp", "127.0.0.1:0")
+		if err != nil {
+			return nil, err
+		}
+		addresses[i] = listener.Addr().String()
+		yarpcInbound := yarpcTransport.NewInbound(listener)
+		yarpcInbound.SetRouter(newTestRouter(procedures))
```

Comment:
what's the argument about just putting all these inbounds into a dispatcher and starting/stopping that once?  Is there sufficient custom functionality here to merit it's own testRouter?

Reply:
The only exposed test router in yarpc-go is one that uses mocks, which I don't want.

Reply:
Oh, you mean a dispatcher as total - I like this current setup without relying on config, ie explicitly setting it up, I don't see too much benefit going to a dispatcher, this gives me more theoretical control.